lib/join/server_env0.go

```diff
@@ -0,0 +1,129 @@
+// Teleport
+// Copyright (C) 2025 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package join
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	workloadidentityv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/workloadidentity/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/join/env0"
+	"github.com/gravitational/teleport/lib/join/internal/authz"
+	"github.com/gravitational/teleport/lib/join/internal/diagnostic"
+	"github.com/gravitational/teleport/lib/join/internal/messages"
+)
+
+type Env0TokenValidator interface {
+	ValidateToken(
+		ctx context.Context,
+		token []byte,
+	) (*env0.IDTokenClaims, error)
+}
+
+// handleTokenJoin handles join attempts for the token join method.
+func (s *Server) handleEnv0Join(
+	stream messages.ServerStream,
+	authCtx *authz.Context,
+	clientInit *messages.ClientInit,
+	provisionToken types.ProvisionToken,
+) (messages.Response, error) {
+	// Receive the TokenInit message from the client.
+	tokenInit, err := messages.RecvRequest[*messages.OIDCInit](stream)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// Set any diagnostic info from the ClientParams.
+	setDiagnosticClientParams(stream.Diagnostic(), &tokenInit.ClientParams)
+
+	verifiedIdentity, err := s.cfg.AuthService.GetEnv0IDTokenValidator().ValidateToken(stream.Context(), tokenInit.IDToken)
+	stream.Diagnostic().Set(func(info *diagnostic.Info) {
+		info.RawJoinAttrs = verifiedIdentity
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "verifying Env0 OIDC token")
+	}
+
+	ptv2, ok := provisionToken.(*types.ProvisionTokenV2)
+	if !ok {
+		return nil, trace.BadParameter("expected *types.ProvisionTokenV2, got %T", provisionToken)
+	}
+
+	if err := checkEnv0AllowRules(ptv2, verifiedIdentity); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// There are no additional checks for the token join method, just make the
+	// result message and return it.
```

Comment:
```suggestion
	// Make and return the final result message it.
```

Reply:
Whoops! I'll just remove the comment outright, it doesn't serve much purpose when it's just doing expected stuff.